crates/wasmparser/src/validator/names.rs

```diff
@@ -547,6 +548,10 @@ impl<'a> ComponentNameParser<'a> {
             return Ok(ParsedComponentNameKind::Hash);
         }
 
+        if self.eat_str("$input") {
+            return Ok(ParsedComponentNameKind::Hash);
+        }
+
```

Comment:
Oh wasm-compose-specific inputs shouldn't make their way into wasmparser. I don't know how `$input` is used much myself, but the fix for this should be in wasm-compose rather than in validation